integration/client/client_windows_test.go

```diff
@@ -22,9 +22,11 @@ import (
 	"path/filepath"
 
 	"github.com/Microsoft/hcsshim/osversion"
-	_ "github.com/Microsoft/hcsshim/test/functional/manifest" // For rsrc_amd64.syso
 )
 
+// Needed to make hscshim APIs work correctly. See #6896
+//go:generate go-winres make --no-suffix --arch amd64 --out rsrc_windows_amd64_test.syso
```

Comment:
> //go:generate go run doesn't work when cross-compiling, because there's no way to make that go run not build a cross-compiled binary that it can't then run without some kind of runner binary.

Is this something we can put in the Makefile instead if `go:generate` is not an option?

Also, do we _need_ to check in the binary .syso files?

Reply:
> Is this something we can put in the Makefile instead if `go:generate` is not an option?
The `go generate` call is made from the Makefile, the problem is that the crossbuild runs the Makefile under the target `GOOS`/`GOARCH`. which is correct because we want to generate the .syso for the target `GOOS`/`GOARCH`.

To be clear, the problem is not `go:generate`, it's using `go run` in `go:generate`. I don't understand the logic of `go run` using the _target_ `GOOS`/`GO ARCH` to build, when its literal job is to execute a binary in the build environment, not the target environment.

> Also, do we need to check in the binary .syso files?

We don't _need_ to, but I'm unaware of a better way to hook the test-executable-building step; otherwise we need to add `go generate ./...` to some top-level in the Makefiles that always runs, or _every entrypoint that might run tests_, and that also means people like me (maybe just me) who're working on Windows without make can't even rely on `go test` in a fresh checkout but need to `go generate` after every checkout/rebase/switch/merge in case something's changed.

My impression from [the introduction of `go:generate`](https://go.dev/blog/generate) and [the linked design document](https://docs.google.com/document/d/1V03LUfjSADDooDMhe-_K59EgpTEm3V8uvQRuNMAEnjg/edit) is that the outputs are expected to be committed, as `go generate` should only be needed when maintaining those generated blobs, not for every user. It's actually unfortunate that the containerd.exe and ctr.exe blobs contain version data, but I suspect we could commit those with the same version data as we use in the `version` package, and in the Makefiles, when `VERSION` is setup, update them in-place (or generate them especially for the release), similarly to the use of the -X flag for LDFlags used to update the `version` package.

However, I only came across the `VERSION` handling at the end of my hack-session, so I haven't really looked hard into that, and particularly, into what `go:generate` and `go run` can do with the command-line. It might almost need a separate tool written, a thin wrapper for the winres library that go-winres thinly wraps, that hard-codes version-calcuations.

I'm really hoping to avoid what moby ended up with, which is a _lot_ of external (and duplicate) scripting to massage things into place, but still achieve the outcome that after a build, your repo isn't all messed up with behaviour-changing junk and modified files.

Reply:
> To be clear, the problem is not `go:generate`, it's using `go run <module>@version` in `go:generate`

Maybe I also wasn't being clear :smile:.  I'm wondering if there's value in using `go:generate` at all given the problem you've outlined, and if instead we could embed the `go-winres` command in question in the Makefile as a dependency for the build & test targets.

> We don't _need_ to, but I'm unaware of a better way to hook the test-executable-building step

Assuming that the Makefile is used to run the tests, hooking in as a dependency in the Makefile seems reasonable to me.

Reply:
"Assuming that the Makefile is used to run the tests" is my concern. That's not neccessary right now, `go test` is sufficient (`make test` and `make root-test` don't do anything else anyway), and would complicate my workflow for running admin-required tests on Windows, since if it's doing anything more than `go test` as an administrator, like installing packages or generating .syso files, it's going to be generating files that will later interfere with normal usage because they were created/owned by administrators. CI can do _everything_ as Administrator on Windows because it's a throwaway VM.

I like `go:generate` because it's _in-source_ documentation of the generated source file, rather than needing to poke around in the build system and the CI yaml, because anything we do in the Makefile for tests is also going to get duplicated around for the non-Makefile parts of CI and the parts of CI that don't hit that script, or use a different Makefile target. You can see this with [how many places I needed to add `go install go-winres` once I could no longer `go run github.com/.../go-winres`](https://github.com/containerd/containerd/pull/6945/commits/2f04ed82819f302fe7489cfd2ae799bb7e34ab46), because even in our GHA CI flow, there's three _unique_ paths to "working env" depending on which test is being run, in order to allow calling `go-winres` (via `go generate`) as part of `make binaries` for the _non-test_ case which is under Makefile control. And I haven't looked, but I suspect the 'nightly' run will also need changing, at least based on how much setup copying I had to do to port the cri testsuite from the nightly run to the GHA CI run.

Reply:
Thanks for clarifying, I really appreciate it.